### ISSUE TITLE
cisco_benigncertain reports false positives

### ISSUE THREAD

**Issue by kunkku** [rapid7/metasploit-framework#9553]
## Steps to reproduce

Use the cisco_benigncertain module on a compliant, non-vulnerable IKE implementation, such as strongSwan.

## Expected behavior

No vulnerability is reported.

## Current behavior

The module reports the target as vulnerable to CVE-2016-6415. The module implementation incorrectly assumes the IKEv1 notification message header to be 36 bytes long, even though the correct length is 40. In IKEv2, it is 36 bytes.

**Comment by snatch2013** [comment]
checked for several cisco IOS and IOS-XE versions, which are not affected by CVE-2016-6415 according to cisco check tool, for example, IOS-XE 16.9.6, IOS 15.7.3M8 , IOS 15.1(4)M12a, but metasploit shows IKE leak.

**Comment by bcoles** [comment]
> checked for several cisco IOS and IOS-XE versions, which are not affected by [CVE-2016-6415](https://github.com/advisories/GHSA-ccm6-q86p-2hwx) according to cisco check tool, for example, IOS-XE 16.9.6, IOS 15.7.3M8 , IOS 15.1(4)M12a, but metasploit shows IKE leak.

Can you show output with `set verbose true` ? This will show the leaked data.

I haven't looked into this, but according to OP the length check is incorrect. The following diff may help:

```diff
diff --git a/modules/auxiliary/scanner/ike/cisco_ike_benigncertain.rb b/modules/auxiliary/scanner/ike/cisco_ike_benigncertain.rb
index 65dfb78326..081945c4b4 100644
--- a/modules/auxiliary/scanner/ike/cisco_ike_benigncertain.rb
+++ b/modules/auxiliary/scanner/ike/cisco_ike_benigncertain.rb
@@ -60,7 +60,7 @@ class MetasploitModule < Msf::Auxiliary
 
       udp_sock.sendto(isakmp_pkt, ip, datastore['RPORT'].to_i)
       res = udp_sock.get(3)
-      return unless res && res.length > 36 # ISAKMP + 36 -> Notitication Data...
+      return unless res && res.length > 40 # ISAKMP + 40 -> Notitication Data...
 
       # Convert non-printable characters to periods
       printable_data = res.gsub(/[^[:print:]]/, '.')
```

**Comment by snatch2013** [comment]
Hello bcoles, 
thanks a lot for your answer. 

For example from IOS 15.7.3M8:
[*] Printable info leaked:
>5..)...:..v[.~>...........(............

[*] Printable info leaked:
>5..)...:..vd..$...........(............

[*] Printable info leaked:
>5..)....."..7$............(............

------------
After applying this fix I don't see any IKE leak from IOS-XE 16.9.6, IOS 15.7.3M8, only from 15.1(4)M12a.
15.1(4)M12a is from 2016, maybe cisco's check tool show incorrect information because this firmware is for 2800 series which is EoS. 

 Thanks a lot for you help!

**Comment by bcoles** [comment]
> After applying this fix I don't see any IKE leak from IOS-XE 16.9.6, IOS 15.7.3M8, only from 15.1(4)M12a.
15.1(4)M12a is from 2016, maybe cisco's check tool show incorrect information because this firmware is for 2800 series which is EoS.

Thanks. I don't trust that Cisco's tool would be 100% accurate either.

Unfortunately without access to test devices it is difficult to provide and test an accurate patch for this.

Presumably the diff above will be useful should someone wish to take on this task.
